File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
#n,m의 위치로 이동할 때 지나야 하는 최소의 칸 수를 구하는 프로그램
from collections import deque

n,m = map(int, input().split())
graph = list(list(map(int, input())) for _ in range(n))

def bfs(x,y):
    queue = deque()
    queue.append((x,y))

    dx = [-1,1,0,0]
    dy = [0,0,-1,1]

    while queue:
        x,y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<=nx<n and 0<=ny<m:
                if graph[nx][ny] == 1:
                    queue.append((nx,ny))
                    graph[nx][ny] = graph[x][y] + 1
                else:
                    continue
    return graph[n-1][m-1]

print(bfs(0,0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.io.*;

public class Main {

    private static int[] dy = {-1, 1, 0, 0};
    private static int[] dx = {0, 0, -1, 1};
    private static int N;
    private static int M;
    private static boolean visited[][];

    public static void main(String[] args) {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        try {
            String[] nums = br.readLine().split(" ");
            N = Integer.parseInt(nums[0]);
            M = Integer.parseInt(nums[1]);
            visited = new boolean[N][M];

            int[][] graph = new int[N][M];

            for (int i = 0; i < N; i++) {
                String line = br.readLine();
                for (int j = 0; j < M; j++) {
                    graph[i][j] = line.charAt(j) - '0';
                }
            }

            bfs(graph, 0, 0);
            System.out.println(graph[N-1][M-1]);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void bfs(int[][] graph, int y, int x) {
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(y, x));
        visited[y][x] = true;

        while (!queue.isEmpty()) {
            Node current = queue.poll();
            int currentY = current.y;
            int currentX = current.x;

            //현재 좌표로 부터 상,하,좌,우 탐색
            for (int i = 0; i < 4; i++) {
                int ny = currentY + dy[i];
                int nx = currentX + dx[i];

                //범위 벗어나는 경우
                if (ny < 0 || nx < 0 || ny >= N || nx >= M) continue;

                //0 인 경우
                if (graph[ny][nx] == 0) continue;

                if (graph[ny][nx] == 1) {
                    visited[ny][nx] = true;
                    graph[ny][nx] = graph[currentY][currentX] + 1;
                    queue.add(new Node(ny, nx));
                }
            }
        }
    }

    static class Node {
        int y;
        int x;

        public Node(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
}
